packages/sdk/src/types/runtime/assets/assets-v2.md

```diff
@@ -0,0 +1,100 @@
+		to be unloaded are not in use during the unload. This wait is done via a new
+		context method that resolves a promise when the next update is done.
+	3. It sends an `unload-assets` message to clients with the asset IDs. This will
+		trigger an `Object.Destroy` in Unity's asset cache, freeing the memory.
+	4. It collects success/failure messages from the client, and propagates any errors
+		from the client unload.
+
+2. `AssetManager.unloadGroup(group: AssetGroup): Promise<void>`
+
+	This simply calls `unloadAssets` with the IDs of all the assets in the group. If
+	alternative two is implemented, this will be the only way to unload container-loaded
+	assets. Even if it isn't, this will be more efficient than piecemeal unloading, but
+	the developer can decide which is more appropriate.
+
+3. `AssetManager.unloadUnused(): Promise<void>`
+
+	Loops over all assets, finds those that have no references, and calls `unloadAssets`
+	with their IDs.
+
+4. `Asset.unload(): Promise<void>`
+
+	Calls `unloadAssets` with the current asset ID.
+
+## Proposed Sync Layer Changes
+
+### Session
+
+Currently the session stores two things related to assets: a list of asset creation
+messages (`load-assets` and `create-asset` payloads), and a map of asset updates
+since creation by ID. This proposal would replace the creation list with a creation
+map indexed by message ID, and would replace the update map with a map of new
+SyncAsset structures. These would contain an asset ID, a creation message ID, a flag
+indicating if the asset has been uploaded, and a cumulative update message. Session
+state would get modified through four functions:
+
+1. `cacheAssetCreationRequest(message: Message<LoadAssets | CreateAsset>): void`
+
+	Store the creation message in the session by message ID.
+
+2. `cacheAssetCreation(assetId: string, creationMessageId: string): void`
+
+	For each asset resulting from a creation message, associate the asset with the
+	creation message in the session.
+
+3. `cacheAssetUpdate(update: Message<AssetUpdate>): void`
+
+	Store the update to an asset in the session, or roll it into the creation message
+	if possible.
+
+4. `cacheAssetUnload(...assetIds: string[]): void`
+
+	Delete the asset mappings and updates from the session, clean out any unused
+	creation messages, and clean out any unloaded SyncAssets still referring to the
+	unused creation message.
+
+### Rules
+
+1. On outgoing `load-assets` and `create-asset` messages, call
+	`session.cacheAssetCreationRequest`.
+2. On incoming `assets-loaded` messages from the authoritative peer, call
+	`session.cacheAssetCreation`.
+3. On outgoing `asset-update` messages, call `session.cacheAssetUpdate`.
+4. On outgoing `unload-assets` messages, call `session.cacheAssetUnload`.
+
+### Client Sync
+
+During the `load-assets` stage of client sync:
+
+1. Send creation messages of all loaded assets.
+2. Send unload messages for all container-loaded assets that have been unloaded.
```

Comment:
Does this mean that if you load and unload an asset 100 of times, it will send 100 load and 100 unload requests? Instead, I'd expect the sync layer to remove the cacheassetcreationrequest/cacheassetcreation/cacheassetupdate for every asset in that assetcontainer/group.

Reply:
Loading a glTF multiple times is not a use case that is currently supported (I hope to support it with this update), but I would expect each load request to produce a duplicate set of assets, not the same set over and over. But to the spirit of your question, if a container contains 100 assets and 99 are unloaded, the sync layer will trigger the load of all 100, then immediately unload 99. If all 100 are unloaded, none get loaded because the whole container got flushed from the cache. So it works as you expect.

---
In reply to: [300189903](https://github.com/microsoft/mixed-reality-extension-sdk/pull/369#discussion_r300189903) [](ancestors = 300189903)

Reply:
If unload is only supported at the container level, then this complexity goes away, correct?

Reply:
This would be a no-op, yes.

---
In reply to: [300767341](https://github.com/microsoft/mixed-reality-extension-sdk/pull/369#discussion_r300767341) [](ancestors = 300767341)